xds/bootstrap/bootstrap.go

```diff
@@ -0,0 +1,65 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package bootstrap provides the functionality to register possible options
+// for aspects of the xDS client through the bootstrap file.
+//
+// Experimental
+//
+// Notice: This package is EXPERIMENTAL and may be changed or removed
+// in a later release.
+package bootstrap
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"google.golang.org/grpc/credentials"
+)
+
+// registry is a map from credential type name to Credential builder.
+var registry = make(map[string]Credentials)
+
+// Credentials interface encapsulates a credential provider
+// that can be used for communicating with the xDS Management server.
+type Credentials interface {
+	// Build returns a credential bundle associated with the Builder.
+	Build(config json.RawMessage) (credentials.Bundle, error)
+	// Name returns the credential name associated with this credential.
+	Name() string
+}
+
+// RegisterCredentials registers the credential builder that can be used to communicate
+// with the xds management server.
+//
+// NOTE: this function must only be called during initialization time (i.e. in
+// an init() function), and is not thread-safe. If multiple credentials are
+// registered with the same name, the one registered last will take effect.
+func RegisterCredentials(c Credentials) {
+	registry[c.Name()] = c
+}
+
+// GetCredentials returns the credentials bundle associated with a given name.
+// If no credentials are registered with the name, an error will be returned.
+func GetCredentials(name string, config json.RawMessage) (credentials.Bundle, error) {
```

Comment:
Return `nil` if not registered? The returned error doesn't have much information.

Reply:
Returning nil, nil is an option, but I think it might be confusing. If the user of the API assumes that if err == nil, the bundle is valid it might lead to a crash. I agree that the error is not very useful, but the alternate option might be more dangerous.

Reply:
No, just return `nil` for the value, and don't have an error return value at all.  That would be my preference, and what we do traditionally.

Reply:
Doug, You suggested the error value because we wouldn't be able to return JSON errors back otherwise. Do we want to go back to the earliest model that I had and ignore possible JSON config issues?

Reply:
Oh, I missed that this function calls `Build()`.
I think it would be OK to return the builder, and wait for the caller to call `Build()`.

Reply:
Done.